_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,292 @@
+* *Latency* measured in the time it takes for messages to be fetched from the broker
+* *Data loss or duplication* when committing offsets or recovering from failure
+* *Handling of transactional messages* from the producer and consumer side
+* *Minimizing the impact of rebalances* to reduce downtime
+
+As with producers, you will want to achieve a balance between throughput and latency that meets your needs. You will also want to ensure your configuration helps reduce delays caused by unnecessary rebalances of consumer groups. Note, however, that you should avoid using any properties that cause conflict with the properties or guarantees provided by your application.
+
+>If you want to read more about what each property does, see Kafka's [consumer configs](https://kafka.apache.org/documentation/#consumerconfigs).
+
+### Message ordering guarantees
+
+Kafka *only* provides ordering guarantees for messages in a single partition. If you want a strict ordering of messages from one topic, the only option is to use one partition per topic. The consumer can then observe messages in the same order that they were committed to the broker.
+
+Potentially, things become less precise when a consumer is consuming messages from multiple partitions. Although the ordering for each partition is kept, the order of messages fetched from _all_ partitions is not guaranteed, as it does not necessarily reflect the order in which they were sent. In practice, however, retaining the order of each partition when consuming from multiple partitions is sufficient for most well-designed consumer client applications.
+
+### Scaling with consumer groups
+
+You'll need a group id if you intend to use consumer groups. Consumer groups are used so commonly, that they might be considered part of a basic consumer configuration. Consumer groups are a way of sharing the work of consuming messages from a set of partitions between a number of consumers by dividing the partitions between them. Consumers are grouped using a `group.id`, allowing messages to be spread across the members that share the same id.
+
+```properties
+# ...
+group.id=my-group-id
+# ...
+```
+
+Consumer groups are very useful for scaling your consumers according to demand. Consumers within a group do not read data from the same partition, but can receive data exclusively from zero or more partitions. One partition is assigned to exactly one member of a consumer group. As mentioned previously, if you want to guarantee message ordering, you would use one consumer with a single partition.
+
+There are a couple of things worth mentioning when thinking about how many consumers to include in a group. Adding more consumers than partitions will not increase throughput. Excess consumers will be partition-free and idle. On the other hand, there is a potential upside in terms of reliability, because those idle consumers will be on standby in the event of failure.
```

Comment:
```suggestion
There are a couple of things worth mentioning when thinking about how many consumers to include in a group. Adding more consumers than partitions will not increase throughput. Excess consumers will be partition-free and idle. This might not be entirely pointless, however, as those an idle consumer is effectively on standby in the event of failure of one of the consumers that does have partitions assigned.
```